[GCP] Minimal permission for service account and instruction for creation (#2112)
skypilot-org/skypilot: docs/source/cloud-setup/cloud-permissions/gcp.rst

.. _cloud-permissions-gcp:

GCP
=============

> 

Reviewer: Generally, the administrator can choose among three "levels" of permissions, from the most permissive and least setup effort, to the least permissive and more setup effort:

- **Default**: no setup, give users Owner-level permissions (i.e., you do not need to follow the instructions here)
- **Medium-level**: easy setup, with a medium set of permissions
- **Minimal**: more setup, with the minimal set of permissions

(Please check if correct)

.. _medium-permissions:

Medium Permissions
-----------------------

The easiest way to grant permissions to a user access your GCP project without the ``Owner`` role is to add the following roles to the user principals:

.. code-block:: yaml

  roles/browser
  roles/compute.admin
  roles/iam.serviceAccountAdmin
  roles/iam.serviceAccountUser
  roles/serviceusage.serviceUsageConsumer
  roles/storage.admin

Optionally, to use TPUs, add the following role:

.. code-block:: yaml

  roles/tpu.admin

You can grant those accesses via the GCP's `IAM & Admin console <https://console.cloud.google.com/iam-admin/iam>`__.

Minimal Permissions
-----------------------

> The :ref:`Medium Permissions <medium-permissions>` assigns strong admin permissions to the user for some GCP services. It is fine for most users, but if you would like to grant fine-grained minimal permissions to your users in your organization / project, you can create a custom role with the permissions required by following the steps below:

Reviewer: ```suggestion
The :ref:`Medium Permissions <medium-permissions>` assigns admin permissions for some GCP services to the user.  If you would like to grant finer-grained and more minimal permissions to your users in your organization / project, you can create a custom role by following the steps below:
```

User
~~~~~~~~~~~~

> 1. Go to the GCP's `IAM & Admin console <https://console.cloud.google.com/iam-admin/roles>`__ and click on **Create Role**.

Reviewer: ```suggestion
1. Go to GCP's `IAM & Admin console <https://console.cloud.google.com/iam-admin/roles>`__ and click on **Create Role**.
```

    compute.firewalls.get
    compute.instances.create 
    compute.instances.delete
    compute.instances.get
    compute.instances.list
    compute.instances.setLabels
    compute.instances.setMetadata
    compute.instances.setServiceAccount
    compute.instances.start
    compute.instances.stop
    compute.networks.get
    compute.networks.list
    compute.networks.getEffectiveFirewalls
    compute.globalOperations.get
    compute.subnetworks.use
    compute.subnetworks.list
    compute.subnetworks.useExternalIp
    compute.projects.get
    compute.projects.setCommonInstanceMetadata
    compute.zoneOperations.get
    iam.roles.get
    iam.serviceAccounts.actAs
    iam.serviceAccounts.get
    serviceusage.services.enable
    serviceusage.services.list
    serviceusage.services.use
    resourcemanager.projects.get
    resourcemanager.projects.getIamPolicy

4. **Optional**: If the user needs to access GCS buckets, you can additionally add the following permissions:

.. code-block:: text

    storage.buckets.create
    storage.buckets.get
    storage.buckets.delete
    storage.objects.create
    storage.objects.delete
    storage.objects.get
    storage.objects.list

> 5. **Optional**: If the user need to access TPU VMs, you can additionally add the following permissions (the following may not be exhaustive, please file an issue if you find any missing permissions):

Reviewer: ```suggestion
5. **Optional**: If the user needs to access TPU VMs, you can additionally add the following permissions (the following may not be exhaustive, please file an issue if you find any missing permissions):
```

.. code-block:: text

    tpu.nodes.create
    tpu.nodes.delete
    tpu.nodes.list
    tpu.nodes.get
    tpu.nodes.update
    tpu.operations.get

5. Click **Create** to create the role.
6. Go back to the "IAM" tab and click on **GRANT ACCESS**.
7. Fill in the email address of the user in the “Add principals” section, and select ``minimal-skypilot-role`` in the “Assign roles” section. Click **Save**.

.. image:: ../../images/screenshots/gcp/create-iam.png
    :width: 80%
    :align: center
    :alt: GCP Grant Access

8. The user should receive an invitation to the project and should be able to setup SkyPilot by following the instructions in :ref:`Installation <installation-gcp>`.

.. note::

>     Similar as the AWS, the user created with the above minimal permissions will not be able to create service accounts to be assigned to the instances. The admin need to follow the instruction :ref:`below <gcp-service-account-creation>` to create a service account shared by all the users in the project.

Reviewer: ```suggestion
    The user created with the above minimal permissions will not be able to create service accounts to be assigned to SkyPilot instances. 
    
    The admin needs to follow the :ref:`instruction below <gcp-service-account-creation>` to create a service account to be shared by all users in the project.
```
